### ISSUE TITLE
Allow plugging in Hibernate ORM's `TypeContributor`, `FunctionContributor`

### ISSUE THREAD

**Issue by yrodiere** [quarkusio/quarkus#32936]
### Description

Those classes are the replacement for `MetadataBuilderContributor`, which is deprecated in Hibernate ORM 6.x.

### Implementation ideas

We could allow inserting these components through settings, `@PersistenceUnitExtension`, or both. See how it's done for `MetadataBuilderContributor` currently.

**Comment by Sanne** [comment]
Thanks for opening this, I had it on my TODO :)

**Comment by sixcorners** [comment]
I use https://github.com/kohsuke/metainf-services to register classes like these fyi. Not sure if it works in graal though.

**Comment by yrodiere** [comment]
Right I suppose that's a workaround.

In the long run, indeed there's the problem of GraalVM, and the fact that "services" probably wouldn't get injected by CDI, which we invariably get a bug report about whenever we allow plugging in a custom component. So I'd favor annotations/configuration backed by CDI instantiation if we are to implement a "supported" solution.

**Comment by Sanne** [comment]
I don't think this would be a problem for GraalVM, as such registrations happen earlier than when the GraalVM restrictions apply: they don't apply to metamodel initialization as that's done during the native-image build, which happens within a regular JVM.

**Comment by IvanPetkov23** [comment]
Hi @yrodiere,
I see custom components can be inserted via settings, but how do we do it using @PersistenceUnitExtension? Any hints? :)

**Comment by yrodiere** [comment]
Hello, and sorry for the delay. Business trips, so this was bad timing :)

If you mean "how to use `TypeContributor` et. al. in a Quarkus application using `@PersistenceUnitExtension`", that's not possible right now, that's the point of this issue, which someone needs to volunteer to address.

If you mean "how do I fix this damn issue already", then first thank you for looking into this, and second I'd recommend checking how it's addressed for other components. Entry points include:

https://github.com/quarkusio/quarkus/blob/25626f31c9f8a1017dd73bc22d9d45d08cd6f92c/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/PersistenceUnitUtil.java#L28

https://github.com/quarkusio/quarkus/blob/250b23fb135a3e3fec60fed33699f08f811c8126/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootEntityManagerFactoryBuilder.java#L207-L211

In the case of `TypeContributor` though, it's very early configuration that needs to be injected in an earlier phase. I think around there should be a good place:

https://github.com/quarkusio/quarkus/blob/250b23fb135a3e3fec60fed33699f08f811c8126/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootMetadataBuilder.java#L189

This "earliness" also means the user bean implementing `TypeContributor` needs to be initialized during static initialization (i.e. for native compilation, during the build). This may or may not be problematic, but the best way to know is to try by adding a test, for example in `integration-tests/jpa-h2`.

**Comment by yrodiere** [comment]
FTR @NickBeginner is planning to work on this.

**Comment by yrodiere** [comment]
Moving `AdditionalMappingContributor` out of scope because it's provided through the serviceloader, which would make its integration pretty different.

It's a very advanced API anyway, mainly useful for Hibernate Envers and Hibernate Search, so let wait until someone has a use case for it.
